src/compiler/binder.ts

```diff
@@ -791,6 +800,13 @@ namespace ts {
                     bindPostfixUnaryExpressionFlow(<PostfixUnaryExpression>node);
                     break;
                 case SyntaxKind.BinaryExpression:
+                    if (isDestructuringAssignment(node)) {

```

Comment:
Doesn't this need to be a (bail) case within `bindBinaryExpressionFlow` to handle when an inner binary expression is a restructuring assignment?

Reply:
I've updated to have `maybeBind` exclude destructuring assignments so that it goes through the regular `bind` flow (and thus hits the escape above).

Reply:
I've also added test cases for the destructuring assignment appearing on either side of a binary expression.